### PR TITLE
Rename all `DataSet` mentions to `Dataset` (excl. docs)

### PR DESCRIPTION
**Pull request by merelcht**
## Description
Solves part of https://github.com/kedro-org/kedro/issues/2129

Will do the renaming in docs separately to limit the number of files changed: https://github.com/kedro-org/kedro/pull/3148

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

File: features/steps/test_starter/{{ cookiecutter.repo_name }}/conf/base/catalog.yml [hunk 1]
```diff
@@ -35,13 +35,13 @@
 #    # if_exists: 'fail'
 #    # if_exists: 'append'
 #
-# The Data Catalog supports being able to reference the same file using two different DataSet implementations
+# The Data Catalog supports being able to reference the same file using two different Dataset implementations
```
Comment (deepyaman) [inline features/steps/test_starter/{{ cookiecutter.repo_name }}/conf/base/catalog.yml:38]:
```suggestion
# The Data Catalog supports being able to reference the same file using two different dataset implementations
```

File: kedro/io/data_catalog.py [hunk 2]
```diff
@@ -176,9 +176,9 @@ def __init__(  # noqa: too-many-arguments
         Example:
         ::
 
-            >>> from kedro_datasets.pandas import CSVDataSet
+            >>> from kedro_datasets.pandas import CSVDataset
```
Comment (deepyaman) [inline kedro/io/data_catalog.py:179]: As commented in the other PR, these are dependent on the `kedro_datasets` version; do we feel 0.19.0 is a good time to just assume they have a new-enough `kedro_datasets`, or should this be communicated more clearly somewhere?

File: kedro/io/partitioned_dataset.py [hunk 3]
```diff
@@ -379,7 +375,7 @@ class IncrementalDataset(PartitionedDataset):
         >>> dataset.load()
     """
 
-    DEFAULT_CHECKPOINT_TYPE = "kedro_datasets.text.TextDataSet"  # TODO: PartitionedDataset should move to kedro-datasets
+    DEFAULT_CHECKPOINT_TYPE = "kedro_datasets.text.TextDataset"  # TODO: PartitionedDataset should move to kedro-datasets
```
Comment (deepyaman) [inline kedro/io/partitioned_dataset.py:378]:
```suggestion
    DEFAULT_CHECKPOINT_TYPE = "kedro_datasets.text.TextDataset"
```
Not sure what this comment is for, even though it's unrelated.

**Review by deepyaman — APPROVED** [review]
LGTM, couple small changes requested

File: kedro/io/partitioned_dataset.py [hunk 4]
```diff
@@ -379,7 +375,7 @@ class IncrementalDataset(PartitionedDataset):
         >>> dataset.load()
     """
 
-    DEFAULT_CHECKPOINT_TYPE = "kedro_datasets.text.TextDataSet"  # TODO: PartitionedDataset should move to kedro-datasets
```
Comment (AhdraMeraliQB) [inline kedro/io/partitioned_dataset.py:382]: Are we not moving PartitionedDataset anymore?

File: kedro/io/partitioned_dataset.py (hunk 3 above)
Comment (AhdraMeraliQB) [reply kedro/io/partitioned_dataset.py]: Are we not moving PartitionedDataset anymore?

Comment (merelcht) [reply kedro/io/partitioned_dataset.py]: Yes we are but we have tickets for that now, much easier to track than a random `TODO` 😅 :
https://github.com/kedro-org/kedro/issues/2713
and https://github.com/kedro-org/kedro-plugins/pull/253, which is ready for review too!

**Review by stichbury — APPROVED** [review]
LGTM 🌟

**Pull request merged**